ambry-router/src/main/java/com.github.ambry.router/GetManager.java

```diff
@@ -111,24 +113,56 @@ public void registerRequestToSend(GetOperation getOperation, RequestInfo request
 
   /**
    * Submit an operation to get a blob asynchronously.
-   * @param blobId The {@link BlobId} associated with the request.
+   * @param blobIdStr The ID of the blob for which blob data is requested.
    * @param options The {@link GetBlobOptionsInternal} associated with the operation.
    * @param callback The {@link Callback} object to be called on completion of the operation.
+   * @throws RouterException if the blobIdStr is invalid.
    */
-  void submitGetBlobOperation(BlobId blobId, GetBlobOptionsInternal options, Callback<GetBlobResultInternal> callback) {
+  void submitGetBlobOperation(String blobIdStr, GetBlobOptionsInternal options,
+      Callback<GetBlobResultInternal> callback) throws RouterException {
     GetOperation getOperation;
+    BlobId blobId = RouterUtils.getBlobIdFromString(blobIdStr, clusterMap);
+    boolean isEncrypted = BlobId.isEncrypted(blobIdStr);
+    if (blobId.getDatacenterId() != ClusterMapUtils.UNKNOWN_DATACENTER_ID
```

Comment:
if you move this within `GetOperation`, then I think the signature of `GetBlobOperation` and `GetBlobInfoOperation` can largely remain unchanged (with only the `BlobId` getting replaced with `String`). You wouldn't need to pass in `isEncrypted`, that can be computed in the constructor of `GetOperation`.

Don't want to drag this patch any further, so do that if you feel it is cleaner.

Reply:
Tried to do this. However `getBlobInfoOperation` constructor needs `operationTracker = getOperationTracker(blobId.getPartition());`(which not needed in `getBlobOperation`). This blocks our way to make blobIdStr -> blobId in `getOperation`.